### PR TITLE
Minor API additions

### PR DESCRIPTION
**Pull request by tnballo**
- RR-related
- New `in_kernel_code_linux`
- Argc/argv wrappers

### CODE REVIEW

File: panda-rs/src/api/rr.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+/// Stop and quit, wraps QMP functions.
+pub fn vm_quit() -> c_int {
+    unsafe {
+        panda_sys::panda_replay_end()
+    }
+}
+
+/// Start recording.
+/// If `snapshot.is_some()` restore the named snapshot prior to recording.
+pub fn record_begin(name: &str, snapshot: Option<&str>) -> Result<c_int, Error> {
```
Comment (jamcleod) [inline panda-rs/src/api/rr.rs:24]: Converting this c_int into a Result itself would probably be preferable, as the int has 3 values: `RRCTRL_EINVALID` (recording already on, `panda_sys::RRCTRL_ret_RRCTRL_EINVALID`), `RRCTRL_EPENDING` (recording in the process of being turned on already, `panda_sys::RRCTRL_ret_RRCTRL_EINVALID`) and `OK` (`panda_sys::RRCTRL_ret_RRCTRL_OK`).

https://github.com/panda-re/panda/blob/ecfe828a302cdd99cf0858063d27624c49dcb3f1/panda/src/callbacks.c#L767-L770

If you don't feel like cramming this in the big error, you could also make an `RrError` enum or something similar.

File: panda-rs/src/api/rr.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                Err(e) => Err(Error::InvalidString(e))
+            },
+            None => Ok(unsafe { panda_sys::panda_record_begin(c_name.as_ptr(), ptr::null()) }),
+        },
+        Err(e) => Err(Error::InvalidString(e))
+    }
+}
+
+/// End currently recording.
+pub fn record_end() -> c_int {
```
Comment (jamcleod) [inline panda-rs/src/api/rr.rs:38]: Similar to the record_begin comment, the following int values could be converted into a Result:

`OK` -> `Ok(())`
`EINVALID` -> recording is not enabled
`EPENDING` -> recording is in the process of being turned off

File: panda-rs/src/api/rr.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        Err(e) => Err(Error::InvalidString(e))
+    }
+}
+
+/// End currently running replay.
+pub fn replay_end() -> c_int {
+    unsafe {
+        panda_sys::panda_replay_end()
+    }
+}
```
Comment (jamcleod) [inline panda-rs/src/api/rr.rs:57]: replay_{begin,end} uses the same error scheme as record_{begin,end}, so you can probably copy/paste the error handling done there.

**Review by jamcleod — COMMENTED** [review]
If you don't feel like implementing my suggestions, feel free to just go ahead and merge and I'll handle them at a future time. If you do implement them, I don't think they'll require any additional review and you can go ahead and merge them after CI passes.

File: panda-rs/src/api/rr.rs (hunk 1 above)
Comment (tnballo) [reply panda-rs/src/api/rr.rs]: Not propagating up `c_int`s is a really good call, thanks!

Went with an `RrError` `enum` that handles translation from `c_int` via an associated method. The existing top-level `Error` now uses `#[error(transparent)]` to pass `RrError` through. Hope this patterns makes it easy to add other "subclasses" of errors in the future while maintaining the same external API!

All relevant return signatures updated from `Result<c_int, Error>` to `Result<(), Error>` where `Ok` == `panda_sys::RRCTRL_ret_RRCTRL_OK`.

File: panda-rs/src/api/rr.rs (hunk 2 above)
Comment (tnballo) [reply panda-rs/src/api/rr.rs]: Done, thanks!

**Comment by tnballo** [comment]
Circled back to this after a while! Review feedback addressed, additional bug in `vm_quit` fixed, CI passed - merging it in.

**Pull request merged**